src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -722,6 +726,57 @@ public void onFileChange(File file) {
         return dockerfileLines;
     }
 
+    private List<String> removeCopyLines(List<String> dockerfileLines, String buildContext) throws PluginExecutionException {
+        List<String> copyLines = new ArrayList<String>();
+        for (String line : dockerfileLines) {
+            // Remove white space from the beginning and end of the line
+            String trimLine = line.trim();
+            if (!trimLine.startsWith("#")) {
+                // Break the Dockerfile line down into segments based on any amount of whitespace.
+                // The command must be to the left of any comments.
+                String[] cmdSegments = trimLine.split("#")[0].split("\\s+");
+                // If the line starts with COPY
+                if (cmdSegments[0].equalsIgnoreCase("COPY")) {
+                    String src = cmdSegments[cmdSegments.length - 2];
+                    String dest = cmdSegments[cmdSegments.length - 1];
+                    if (validateSrcMount(src, buildContext)) {
+                        srcMount.add(buildContext + "/" + src);
+                        destMount.add(formatDestMount(dest, src));
+                        copyLines.add(line);
+                    }
+                }
+            }
+        }
+        debug("COPY lines: " + copyLines.toString());
+        dockerfileLines.removeAll(copyLines);
+        return dockerfileLines;
+    }
+
+    private boolean validateSrcMount(String srcMountString, String buildContext) throws PluginExecutionException {
+        File srcMountFile = new File(buildContext + "/" + srcMountString);
+        if (srcMountFile.isDirectory()) {
+            warn("Files in the directory " + srcMountFile + " will not be able to be hot deployed for the dev mode container. " + 
+                "To allow files to be hot deployed, specify individual files when using the COPY command in your Dockerfile");
+            return false;
+        }
+        else if (!srcMountFile.exists()) {
+            error(srcMountFile + " does not exist and will not be copied to the image.");
+            throw new PluginExecutionException("Cannot build docker image with missing file " + srcMountFile);
+        }
+        return true;
+    }
+
+    private String formatDestMount(String destMountString, String srcMountString) {
+        // Cannot mount a file onto a directory, so must add a filename to the end of the destination argument for mounting
+        if (destMountString.endsWith("/") || destMountString.endsWith("\\")) {
+            // Split on forward and backward slashes
+            String [] segments = srcMountString.split("/|\\\\");
+            destMountString += segments[segments.length -1];
```

Comment:
Would it be simpler to get a File or Path object from the source mount, and get the filename from that object?

Although the COPY source might contain wildcards according to  https://docs.docker.com/engine/reference/builder/#copy (should have that as a separate issue)

Reply:
I can try that method out and see how it goes.

As for the wildcards, I agree we need to handle that. We also have a few other scenarios to consider as well for both WAR file and COPY lines.